src/server/game/Entities/Player/Player.cpp

```diff
@@ -27098,6 +27138,43 @@ bool Player::CanSeeSpellClickOn(Creature const* c) const
     return false;
 }
 
+bool Player::CanSeeGossipOn(Creature const* creature) const
+{
+    if (creature->HasNpcFlag(UNIT_NPC_FLAG_GOSSIP))
+    {
+        if (creature->GetCreatureTemplate()->GossipMenuIds.size() || creature->GetGossipMenuId())
+        {
+            if (GetGossipMenuForSource(creature))
+                return true;
+        }
+    }
+
+    // for cases with questgiver/ender without gossip menus
+    if (creature->HasNpcFlag(UNIT_NPC_FLAG_QUESTGIVER))
+    {
+        QuestRelationResult objectQR = sObjectMgr->GetCreatureQuestRelations(creature->GetEntry());
```

Comment:
Move this to just before loop (so that its never called if any of objectQIR returns true)